distribution/lib/Standard/Base/0.0.0-dev/src/Data/Regression.enso

```diff
@@ -0,0 +1,112 @@
+            Fitted_Logarithmic_Model fitted.slope fitted.intercept r_squared
+        Power_Model ->
+            log_xs = ln_series known_xs "X-values"
+            log_ys = ln_series known_ys "Y-values"
+            fitted = Regression.fit_linear log_xs.to_array log_ys.to_array
+            model = Fitted_Power_Model fitted.intercept.exp fitted.slope
+            r_squared = known_ys.compute (Statistics.R_Squared (known_xs.map model.predict))
+            Fitted_Power_Model fitted.intercept.exp fitted.slope r_squared
+        _ -> Error.throw (Illegal_Argument_Error "Unsupported model")
+
+type Fitted_Model
+    ## Fitted line (y = slope x + intercept).
+    type Fitted_Linear_Model slope:Number intercept:Number r_squared:Number=0.0
+
+    ## Fitted exponential line (y = a exp(b x)).
+    type Fitted_Exponential_Model a:Number b:Number r_squared:Number=0.0
+
+    ## Fitted logarithmic line (y = a log x + b).
+    type Fitted_Logarithmic_Model a:Number b:Number r_squared:Number=0.0
+
+    ## Fitted power series (y = a x ^ b).
+    type Fitted_Power_Model a:Number b:Number r_squared:Number=0.0
+
+    ## Display the fitted line.
+    to_text : Text
+    to_text = case self of
+        Fitted_Linear_Model slope intercept _ -> slope.to_text + " * X + " + intercept.to_text
+        Fitted_Exponential_Model a b _ -> a.to_text + " * (" + b.to_text + " * X).exp"
+        Fitted_Logarithmic_Model a b _ -> a.to_text + " * X.ln + " + b.to_text
+        Fitted_Power_Model a b _ -> a.to_text + " * X ^ " + b.to_text
+
+    ## Use the model to predict a value.
+    predict : Number -> Number
+    predict x = case self of
+        Fitted_Linear_Model slope intercept _ -> slope * x + intercept
+        Fitted_Exponential_Model a b _ -> a * (b * x).exp
+        Fitted_Logarithmic_Model a b _ -> a * x.ln + b
+        Fitted_Power_Model a b _ -> a * x ^ b
+        _ -> Error.throw (Illegal_Argument_Error "Unsupported model")
+
+## PRIVATE
+
+   Computes the natural log series as long as all values are positive.
+ln_series : Vector -> Vector ! Illegal_Argument_Error
+ln_series xs series_name="Values" =
+    error_throw p = Error.throw p
+    ln_with_panic x = if x.is_nothing then Nothing else
+        if x <= 0 then Panic.throw (Illegal_Argument_Error (series_name + " must be positive.")) else x.ln
+    Panic.catch Illegal_Argument_Error handler=error_throw <| xs.map ln_with_panic
+
+## PRIVATE
+
+   An error thrown when the linear regression cannot be computed.
+
+   Arguments:
+   - message: The error message.
+type Fit_Error message
+
+## PRIVATE
+
+   Converts the `Fit_Error` to a human-readable representation.
+Fit_Error.to_display_text : Text
+Fit_Error.to_display_text = "Could not fit the model: " + self.message.to_text
+
+## PRIVATE
+Fit_Error.handle_java_exception =
+    throw_fit_error caught_panic =
+        cause = caught_panic.payload.cause
+        Error.throw (Fit_Error cause.getMessage)
+    Panic.catch FitError handler=throw_fit_error
```

Comment:
`catch_java` is made exactly for situations like these so I'd suggest using that:

```suggestion
    Panic.catch_java FitError handler=(java_exception-> Error.throw (Fit_Error java_exception.getMessage))
```

Reply:
This is becoming a pretty common pattern, i.e.:
```
handle_java_exception java_type enso_constructor =
    Panic.catch_java java_type handler=(java_exception-> Error.throw (enso_constructor java_exception.getMessage))
```

I'm wondering if we should encapsulate it into some helper function. OTOH the pattern is very simple and I'm not sure if having the explicit code isn't clearer than replacing it with a helper after all...

Reply:
Think one to revisit when Polyglot_Error is overhauled! The one-line approach is nice enough.